python/dolfinx/fem/petsc.py

```diff
@@ -1118,3 +1121,342 @@ def interpolation_matrix(space0: _FunctionSpace, space1: _FunctionSpace) -> PETS
+
+        self._L = _create_form(
+            F, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+        if J is None:
+            V = u.function_space
+            du = ufl.TrialFunction(V)
+            J = ufl.derivative(F, u, du)
+        self._a_prec = (
+            None
+            if P is None
+            else _create_form(
+                P, form_compiler_options=form_compiler_options, jit_options=jit_options
+            )
+        )
+        self._a = _create_form(
+            J, form_compiler_options=form_compiler_options, jit_options=jit_options
+        )
+
+        self.bcs = bcs
+        self.u = u
+
+    @property
+    def L(self) -> Form:
+        """Compiled linear form of the residual `F(u, v)`"""
+        return self._L
+
+    @property
+    def a(self) -> Form:
+        """Compiled bilinear form of the Jacobian `J`"""
+        return self._a
+
+    @property
+    def P(self) -> Form:
+        """Compiled bilinear form of the preconditioner `P`"""
+        return self._a_prec
+
+    def replace_solution(self, x: PETSc.Vec):
+        """Update the solution for the unknown `u` with the values in `x`.
+
+        Args:
+            x: Data that is inserted into the solution vector
+        """
+        x.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+        x.copy(self.u.x.petsc_vec)
+        self.u.x.petsc_vec.ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)
+
+    def copy_solution(self, x: PETSc.Vec):
+        """Copy the data in `u` into the vector `x`.
+
+        Args:
+            x: Vector to insert data into
+        """
+        self.u.x.petsc_vec.copy(x)
+
+    def F(self, snes: PETSc.SNES, x: PETSc.Vec, F: PETSc.Vec):
+        """Assemble the residual into the vector `F`.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the point to evaluate the residual at.
+            F: Vector to assemble the residual into
+        """
+        self.replace_solution(x)
+        with F.localForm() as f_local:
+            f_local.set(0.0)
+        dolfinx.fem.petsc.assemble_vector(F, self.L)
+        dolfinx.fem.petsc.apply_lifting(F, [self.a], bcs=[self.bcs], x0=[x], alpha=-1.0)
+        F.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+        dolfinx.fem.petsc.set_bc(F, self.bcs, x, -1.0)
+
+    def J(self, snes, x, J, P):
+        """Assemble the Jacobian matrix and preconditioner.
+
+        Args:
+            snes: The solver instance
+            x: The vector containing the point to evaluate at
+            J: Matrix to assemble the Jacobian into
+            P: Matrix to assemble the preconditioner into
+        """
+        # Copy existing soultion into the function used in the residual and Jacobian
+        self.replace_solution(x)
+
+        # Assemble Jacobian
+        J.zeroEntries()
+        dolfinx.fem.petsc.assemble_matrix(J, self.a, self.bcs)
```

Comment:
For consistency you may want to add a `diagonal=1.0` even here, since you have it the call below.